File: src/model/ModelRunner.js
global.acorn = require('../../third-party/acorn')
const JSInterpreter = require('../../third-party/interpreter')
const MousePlugin = require('../plugins/MousePlugin')
const KeyboardPlugin = require('../plugins/KeyboardPlugin')
const ScreenPlugin = require('../plugins/ScreenPlugin')
const SystemPlugin = require('../plugins/SystemPlugin')
const FilePlugin = require('../plugins/FilePlugin')
const NodeTypes = require('./steps/StepTypes')

class ModelRunner {

  constructor () {
    this.initInterpreter()
  }

  execute (code) {
    this._interpreter.appendCode(code)
    this._interpreter.run()
    return this._interpreter.value
  }

  getVariable (name) {
    const scope = this._interpreter.getGlobalScope()
    return scope.object.properties[name]
  }

  initInterpreter () {
    this._interpreter = new JSInterpreter('', function (interpreter, globalObject) {
      new MousePlugin(interpreter, globalObject).init()
      new KeyboardPlugin(interpreter, globalObject).init()
      new ScreenPlugin(interpreter, globalObject).init()
      new SystemPlugin(interpreter, globalObject).init()
      new FilePlugin(interpreter, globalObject).init()
    })
  }

  isFinished (step) {
    return (step === undefined) || step.type === NodeTypes.STOP
  }

  async run (model) {
    let currentStep = model.findEntryPoint().getNextStep()
    while (!this.isFinished(currentStep)) {
      currentStep = await currentStep.execute(this)
    }
  }

  setVariable (varName, value) {
    const scope = this._interpreter.getGlobalScope()
    scope.object.properties[varName] = value
  }
}

module.exports = ModelRunner

File: Models/ModelRunner.cs
using Microsoft.CodeAnalysis.CSharp.Scripting;
using Microsoft.CodeAnalysis.Scripting;
using WorkflowEngine.Models.Steps;
using WorkflowEngine.Plugins;

namespace WorkflowEngine.Models;

public class ModelRunner
{
    private readonly FilePlugin _filePlugin = new();
    private readonly KeyboardPlugin _keyboardPlugin = new();
    private readonly MousePlugin _mousePlugin = new();
    private readonly ScreenPlugin _screenPlugin = new();
    private readonly SystemPlugin _systemPlugin = new();
    private readonly Dictionary<string, object?> _variables = new();

    public async Task<object?> ExecuteAsync(string code)
    {
        try
        {
            var options = ScriptOptions.Default
                .WithReferences(typeof(Console).Assembly)
                .WithImports("System", "System.Console", "System.IO", "System.Threading.Tasks");

            var globals = new ScriptGlobals
            {
                Variables = _variables,
                Mouse = _mousePlugin,
                Keyboard = _keyboardPlugin,
                Screen = _screenPlugin,
                System = _systemPlugin,
                File = _filePlugin
            };

            var result = await CSharpScript.EvaluateAsync(code, options, globals);
            return result;
        }
        catch (Exception ex)
        {
            throw new Exception($"執行腳本時發生錯誤: {ex.Message}");
        }
    }

    public object? GetVariable(string name)
    {
        return _variables.TryGetValue(name, out var value) ? value : null;
    }

    public void SetVariable(string name, object? value)
    {
        _variables[name] = value;
    }

    public async Task RunAsync(Model model)
    {
        var currentStep = model.FindEntryPoint().GetNextStep();

        while (!IsFinished(currentStep))
        {
            currentStep = await currentStep.ExecuteAsync(this);
        }
    }

    private static bool IsFinished(Step? step)
    {
        return step == null || step.Type == StepTypes.STOP;
    }

    public class ScriptGlobals
    {
        public Dictionary<string, object?> Variables { get; set; } = new();
        public MousePlugin Mouse { get; set; } = new();
        public KeyboardPlugin Keyboard { get; set; } = new();
        public ScreenPlugin Screen { get; set; } = new();
        public SystemPlugin System { get; set; } = new();
        public FilePlugin File { get; set; } = new();
    }
}